Deprecate unnamed foreign key constraints in `$droppedForeignKeys` of `TableDiff`

santysisi opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | bug
| Fixed issues | #7124 

#### Summary

This PR introduces a deprecation notice when a unnamed foreign key constraints in `$droppedForeignKeys` is passed in the constructor of  `TableDiff`.

santysisi commented:
I added the test to `SchemaManagerFunctionalTestCase` instead of `SQLiteSchemaManagerTest` because I believe it's relevant for all platforms, not just SQLite.
However, if this test should be scoped to SQLite only, please let me know and I’ll move it accordingly.

morozov commented on src/Platforms/SQLitePlatform.php:
```diff
@@ -929,8 +930,8 @@ private function getForeignKeysInAlteredTable(TableDiff $diff): array
 
-        foreach ($diff->getDroppedForeignKeys() as $constraint) {
-            $constraintName = $constraint->getName();
+        foreach ($diff->getDroppedForeignKeys() as $name => $constraint) {
```
This is against the effort made in https://github.com/doctrine/dbal/pull/7102. The constraint name is represented by the constraint object itself, not by some array key.

santysisi replied:
Got it, thanks for pointing that out! I’ll look into applying a different approach that aligns better with the changes from #7102 as soon as possible.

morozov replied:
It might be a good idea to just deprecate the scenario when the constraint being dropped doesn't have a name. Currently, we're not doing anything, which causes a bug. Making it work correctly would require a deeper redesign of schema management.

Just for context, the fundamental flaw is the following. The test attempts to perform a schema migration in the following steps:
1. Define two versions of a table.
2. Calculate a difference between them.
3. Apply the difference.

The problem is that:
1. The calculation of the difference is non-deterministic (you cannot reliably tell whether a given column was renamed or dropped and created). This doesn't apply to this specific case, but still.
2. SQLite doesn't support `ALTER TABLE`. The DBAL has to recreate it from scratch and migrate the data.

The combination of these flaws leads to the fact that we cannot migrate schemas with unnamed constraints. A proper design for SQLite would not involve the diff at all. For SQLite, it would use the expected table schema directly.

santysisi replied:
Hi, thanks a lot for the detailed explanation! 😊
That makes sense. I’ll update the PR this weekend to only emit a deprecation notice in this case.

Thanks again! ❤️

santysisi committed:
Deprecate unnamed foreign key constraints in `$droppedForeignKeys` of `TableDiff`
DroppedForeignKey must have a constraint name to ensure reliable
schema diffing and migration generation. A deprecation warning
is now triggered if the name is missing.

santysisi commented:
Hi! 😄
I've updated the PR to trigger a deprecation. Additionally, I've changed the base branch, as well as the PR title and description.

santysisi commented:
Hi @morozov, I hope you're doing well! 😄

I've updated this PR based on your suggestions from the previous one. I modified the deprecation message, updated the `UPGRADE.md`, and removed unnecessary code in the test.

Please let me know if anything needs to be changed or corrected 😊

Thanks again! ❤️

morozov commented on UPGRADE.md:
```diff
@@ -8,6 +8,10 @@ awareness about deprecated code.
 
+## Deprecated dropping unnamed constraints in SQLite
+
+Dropping a constraint without specifying its name has been deprecated on the SQLite platform.
```
This will have to be reworded accordingly. Something like this:

> Passing unnamed foreign key constraints as part of the `$droppedForeignKeys` argument of the `TableDiff` constructor has been deprecated.

morozov commented on src/Platforms/SQLitePlatform.php:
```diff
@@ -934,6 +934,11 @@ private function getForeignKeysInAlteredTable(TableDiff $diff): array
             $constraintName = $constraint->getName();
 
             if ($constraintName === '') {
+                Deprecation::trigger(
```
Whether a foreign key constraint is dropped on SQLite or any other platform, it's expected to have a name: https://github.com/doctrine/dbal/blob/aa6b906fcf0cf09e3adde1c623131accd55de63b/src/Platforms/AbstractPlatform.php#L1372-L1374

The fact that `TableDiff::getDroppedForeignKeys()` can return unnamed constraints is essentially a bug as using it as is will result in invalid DDL.

I think we should move this deprecation to the `TableDiff` constructor. In 5.0.x, this deprecation will need to be promoted to an exception, and all callers of `TableDiff::getDroppedForeignKeys()` could make an assertion that the constraint being dropped does have a name before using it.

santysisi replied:
Hi! 😄
Thanks so much for the suggestions ❤️

I’ve made the following changes in the test:

* Updated the test a bit, it now verifies that the deprecation is triggered in the constructor of `TableDiff`.
* Also, it checks that a foreign key is not dropped when the dropped foreign key doesn't have a name.
* Moved the test from `SQLiteSchemaManagerTest` to `SchemaManagerFunctionalTestCase` so it runs across all platforms.

I do have one question. I’m still getting familiar with Doctrine DBAL 😅

Do you think this check is necessary?

```php
if ($constraintName === null || $constraintName === '') {
```

Or would just checking `$constraintName === null` be enough? 🤔

Thanks again for your help! 😊

morozov commented on src/Platforms/AbstractPlatform.php:
```diff
@@ -1370,6 +1370,11 @@ protected function getPreAlterTableIndexForeignKeySQL(TableDiff $diff): array
         foreach ($diff->getDroppedForeignKeys() as $foreignKey) {
+            $constraintName = $foreignKey->getObjectName()?->toString();
+            if ($constraintName === null || $constraintName === '') {
+                continue;
```
We don't want to do this. It is better to have invalid DDL generated, so the user sees that something is wrong, rather then get an invalid input silently ignored.

morozov commented on src/Schema/TableDiff.php:
```diff
@@ -56,6 +56,18 @@ public function __construct(
         }
 
+        foreach ($droppedForeignKeys as $droppedForeignKey) {
+            $constraintName = $droppedForeignKey->getObjectName()?->toString();
+            if ($constraintName === null || $constraintName === '') {
```
We cannot use `getObjectName()` in the library code in 4.4.x since it can throw an exception, if the name is invalid and cannot be parsed using the 5.0.x logic, but can be using the 4.4.x one. Please use `if ($droppedForeignKey->getName() === '')` instead.

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -444,6 +447,76 @@ public function testListTableColumnsWithFixedStringColumn(): void
         self::assertSame(2, $columns['column_char']->getLength());
     }
 
+    public function testAlterTableWithUnnamedForeignKey(): void
```
Please replace this with a `TableDiff` constructor test.

morozov commented on tests/Schema/TableDiffTest.php:
```diff
@@ -0,0 +1,40 @@
+            ->create();
+
+        $droppedForeignKeys = ForeignKeyConstraint::editor()
+            ->setUnquotedReferencingColumnNames('c1', 'c2')
```
`c2` can be removed from the definition: it's not needed to reproduce the issue and it doesn't exist on the referencing table.

morozov commented:
Thanks, @santysisi!

andersonamuller commented on UPGRADE.md:
```diff
@@ -8,6 +8,11 @@ awareness about deprecated code.
 
 # Upgrade to 4.4
 
+## Deprecated dropping unnamed constraints in SQLite
```
@santysisi Does it still only apply for SQLite?

```suggestion
## Deprecated dropping unnamed constraints
```

santysisi replied:
You're right, that was my mistake.
Thanks so much for catching that! 🙏
I've opened this [PR](https://github.com/doctrine/dbal/pull/7155) to resolve the issue
